### DISCUSSION TITLE
Add MongoDB client promise for usage in Auth providers

### DISCUSSION THREAD

**Discussion by yandodov** [payloadcms/payload#7757] [Feature Requests & Ideas]
I'm using Payload `3.0.0-beta.84` with Auth.js and its [MongoDB adapter](https://authjs.dev/getting-started/adapters/mongodb) to persist the sessions in the database.

The problem is that I have to provide the connected MongoDB client (or a promise for it) to Auth.js and Payload doesn't provide it. This means I have to do nasty workarounds with exporting promises:

`src/payload.config.ts`

```ts
import { buildConfig } from "payload";
import { resolvePromise } from "./db";

export default buildConfig({
  // ...
  onInit(payload) {
    resolvePromise(payload.db.connection.getClient());
  },
});
```

`src/auth.ts`

```ts
import { MongoDBAdapter } from "@auth/mongodb-adapter";
import NextAuth from "next-auth";
import Okta from "next-auth/providers/okta";
import { clientPromise } from "./db";

export const { handlers, signIn, signOut, auth } = NextAuth({
  adapter: MongoDBAdapter(clientPromise),
  providers: [Okta],
});
```

`src/db.ts`

```ts
import type payload from 'payload'

type MongoClient = ReturnType<typeof payload.db.connection.getClient>

export let clientPromise: Promise<MongoClient>
export let resolvePromise: (value: MongoClient) => void

if (process.env.NODE_ENV === 'development') {
  // In development mode, use a global variable so that the value
  // is preserved across module reloads caused by HMR (Hot Module Replacement).
  let globalWithMongo = global as typeof globalThis & {
    _mongoClient?: typeof clientPromise
  }
  if (!globalWithMongo._mongoClient) {
    globalWithMongo._mongoClient = new Promise<MongoClient>((resolve) => (resolvePromise = resolve))
  }
  clientPromise = globalWithMongo._mongoClient
} else {
  // In production mode, it's best to not use a global variable.
  clientPromise = new Promise<MongoClient>((resolve) => (resolvePromise = resolve))
}
```

# What I tried

If I try this straight away:

```ts
export const { handlers, signIn, signOut, auth } = NextAuth({
  adapter: MongoDBAdapter(payload.db.connection.getClient()),
  // ...
})
```

…I get the following error:

> TypeError: Cannot read properties of undefined (reading 'connection')

…so the `db` object is not available yet.

# Ideal solution

If Payload has a promise that resolves when the MongoDB client is available, I could simply do:

```ts
export const { handlers, signIn, signOut, auth } = NextAuth({
  adapter: MongoDBAdapter(payload.dbClientPromise),
  // ...
})
```

**Comment by DanRibbens** [comment]
We have this for the postres and sqlite adapters. You could make a PR to add it to mongodb.

It is named ` initializing` and we have `resolveInitializing` and `rejectInitializing`. So in your case you'd have `payload.db.initializing`, 

In the db type we have: 
`initializing: Promise<void>`
`resolveInitializing: () => void`, and 
`rejectInitializing: () => void`

The adapters set it up in `index.ts` with:
```
    let resolveInitializing
    let rejectInitializing

    const initializing = new Promise<void>((res, rej) => {
      resolveInitializing = res
      rejectInitializing = rej
    })
```

In connect, we call it:
`if (typeof this.resolveInitializing === 'function') this.resolveInitializing()` andon error connecting: `if (typeof this.rejectInitializing === 'function') this.rejectInitializing()`

and destroy resets it: 
```
  this.initializing = new Promise((res, rej) => {
    this.resolveInitializing = res
    this.rejectInitializing = rej
  })
```

**Reply by yandodov** [reply]
Thanks for the speedy reaction! I've implemented the feature on branch [feat/mongodb-client-promise](https://github.com/hdodov/payload/tree/feat/mongodb-client-promise) in my fork.

However, that still doesn't quite fix my issue. If you look at my initial comment, `payload.db` is `undefined`, so `payload.db.initializing` is undefined as well.

I have a [reproduction branch](https://github.com/hdodov/payload/tree/feat/mongodb-client-promise-issue) too. Here's how to see the issue:

1. Clone the repo and check out the branch
2. Run `pnpm dev admin-root`
3. Open `http://localhost:3000/my-route`
4. You'll see the error as described here https://github.com/hdodov/payload/commit/194ae1a44541e12a58c4a5b602c979d023f59a70:

   > TypeError: Cannot read properties of undefined (reading 'initializing')

---

I think it could work if we have `payload.dbInitializing` because the actual `payload` object itself isn't undefined. What do you think?
